Change unchecked unwrapping to unwrap_or_default for Args-trait (#1845)

ChrisRega opened the pull request:
Simply changing the unwrap to unwrap_or_default in order to keep the number of arguments identical to the original iterator contents.

sylvestre commented:
Could you please add a test verifying that the old issue doesn't happen anymore?
thanks

ChrisRega commented:
> Could you please add a test verifying that the old issue doesn't happen anymore?
> thanks

Added unit-testing for the desired behaviour.

Arcterus commented:
I’m not sure this is ideal, given that it silently ignores non-UTF-8 arguments.

ChrisRega commented:
> I’m not sure this is ideal, given that it silently ignores non-UTF-8 arguments.

We could either go for some eprintln-warning that an argument was ignored or make this function return a result itself and make the client-code deal with it. There are around 50 usages of it, most look like they collect the arguments before giving it to cl argument parsing. Dealing with results there would produce a lot of very similar code, which then has to be maintained. I see two possible solutions:

1. Ignore the arguments that have broken encoding (with or without warning) and try to continue the program
2. Abort execution alltogether since the user probably wanted to do something with this argument and we clearly can't interpret his intention.

Now that I think a bit more about it, option 2 seems even more sane than just ignoring a parameter (even with a warning).

sylvestre commented:
Would it be possible to add tests at the binary leve? For example in https://github.com/uutils/coreutils/blob/master/tests/by-util/test_basename.rs ?
thanks

ChrisRega commented:
> Would it be possible to add tests at the binary leve? For example in https://github.com/uutils/coreutils/blob/master/tests/by-util/test_basename.rs ?
> thanks

Absolutely, will implement some tomorrow! Thanks for all the feedback

ChrisRega commented:
> > Would it be possible to add tests at the binary leve? For example in https://github.com/uutils/coreutils/blob/master/tests/by-util/test_basename.rs ?
> > thanks
> 
> Absolutely, will implement some tomorrow! Thanks for all the feedback

Turns out, that implementing a test for basename is not trivial after all. UCommand parses its arguments with unwrap directly without ever calling collect_str.
The next problem is with the result of the test, basename with a broken argument should fail like with no argument. It does not, so i thought about changing the behaviour of collect_str to drop items with broken encoding. But this causes more problems in other tests. I will try to investigate this issue a bit more in the next days. I am not satisfied with the current behaviour.

sylvestre reviewed (changes requested):
OK, please let me know when it is ready to be reviewed!

ChrisRega commented on src/uucore/src/lib/lib.rs:
```diff
@@ -67,10 +67,63 @@ pub use crate::features::wide;
-    fn collect_str(self) -> Vec<String> {
-        // FIXME: avoid unwrap()
-        self.map(|s| s.into_string().unwrap()).collect()
+    /// Converts each iterator to a String and collects these into a vector
+    /// On failure, the result will may be lossy or incomplete
+    /// # Arguments
+    /// * `handling` - This switch allows to define the behavior, when invalid encoding is encountered
+    fn collect_str(self, handling: InvalidEncodingHandling) -> ConversionResult {
+        let mut full_conversion = true;
+        let result_vector: Vec<String> = self
+            .map(|s| match s.into_string() {
+                Ok(string) => string,
+                Err(s_ret) => {
+                    full_conversion = false;
+                    let lossy_conversion = s_ret.to_string_lossy();
+                    eprintln!(
+                        "Input with broken encoding occured! (s = '{}') ",
+                        &lossy_conversion
+                    );
+                    match handling {
+                        InvalidEncodingHandling::Ignore => String::new(),
+                        InvalidEncodingHandling::ConvertLossy => lossy_conversion.to_string(),
+                        InvalidEncodingHandling::Panic => {
+                            panic!("Broken encoding found but caller cannot handle it")
+                        }
+                    }
+                }
+            })
+            .collect();
+        match full_conversion {
+            true => ConversionResult::FullyDecoded(result_vector),
+            false => ConversionResult::Lossy(result_vector),
+        }
+    }
+
+    /// convience function for a more slim interface
+    fn collect_str_lossy(self) -> ConversionResult {
+        collect_str(InvalidEncodingHandling::ConvertLossy)
     }
```
@sylvestre what do you think of this? It allows a slim interface with convenience functions like collect_str_lossy while still giving the callers control of what should happen in cases of error. Also, as a caller, you will get a result-like type which allows you to react. There are two example-usages: head and dirname. If you like the design, i will finish this by adding testing and changing all callers to the new API (i would figure the desired behaviour of each command to match the original coreutils).

sylvestre commented:
@Arcterus what do you think about this change? Thanks

ChrisRega commented:
> @Arcterus what do you think about this change? Thanks

I just added unit testing for the conversion. Feedback on this would be appreciated :)

sylvestre commented:
There are some conflicts :)

Besides that, I am fine but I would like @Arcterus to review it

ChrisRega commented:
> There are some conflicts :)
> 
> Besides that, I am fine but I would like @Arcterus to review it

Merged master, fixed the conflicts and updated the pr :)

sylvestre commented:
Looks like it is conflicting again, sorry :/

ChrisRega commented:
> Looks like it is conflicting again, sorry :/

Merged master in again and fixed some hiccups

sylvestre commented:
Sorry but you still have build failures:
```

error[E0433]: failed to resolve: use of undeclared type or module `OsString`
Error:    --> D:\a\coreutils\coreutils\tests\by-util\test_basename.rs:115:21
    |
115 |       let os_string = OsString::from_wide(&source[..]);
    |                       ^^^^^^^^
    |
help: a struct with a similar name exists
    |
115 |     let os_string = String::from_wide(&source[..]);
    |                     ^^^^^^
help: consider importing one of these items
    |
1   | use crate::test_sync::fs::OsString;
    |
1   | use std::ffi::OsString;
```

sylvestre commented on tests/by-util/test_basename.rs:
```diff
@@ -84,3 +85,23 @@ fn test_no_args() {
     expect_error(vec!["a", "b", "c"]);
 }
+
+fn test_invalid_utf8_args(os_str: &OsStr) {
```
Missing
```suggestion
#[test] 
fn test_invalid_utf8_args(os_str: &OsStr) {
```

sylvestre commented on src/uucore/src/lib/lib.rs:
```diff
@@ -84,3 +164,79 @@ pub fn args() -> impl Iterator<Item = String> {
+        );
+    }
+
+    fn test_invalid_utf8_args_ignore(os_str: &OsStr) {
```
#test missing ?

ChrisRega replied:
See above comment

sylvestre commented on src/uu/unlink/src/unlink.rs:
```diff
@@ -17,12 +17,13 @@ use libc::{lstat, stat, unlink};
 
 pub fn uumain(args: impl uucore::Args) -> i32 {
-    let args = args.collect_str();
+    let args = args.collect_str(InvalidEncodingHandling::ConvertLossy).accept_any();
```
rustfmt isn't happy here :)

ChrisRega replied:
That's bad, must have gone wrong with the conflict resolution - i will fix it asap.

sylvestre replied:
np

sylvestre reviewed (changes requested):
some nits
but I would appreciate also a more detailled description  of the why (less technical)

sylvestre commented on tests/by-util/test_basename.rs:
```diff
@@ -84,3 +85,23 @@ fn test_no_args() {
     expect_error(vec!["a", "b", "c"]);
 }
+
+fn test_invalid_utf8_args(os_str: &OsStr) {
```
Missing #[test] ?

ChrisRega replied:
called by line 242: 
`#[cfg(any(unix, target_os = "redox"))]
    #[test]
    fn invalid_utf8_args_unix() {
        use std::os::unix::ffi::OsStrExt;

        let source = [0x66, 0x6f, 0x80, 0x6f];
        let os_str = OsStr::from_bytes(&source[..]);
        test_invalid_utf8_args_lossy(os_str);
        test_invalid_utf8_args_ignore(os_str);
    }`
Allows for no code duplication with illegal windows wide-string encoding (not implemented yet due to missing windows machine)

ChrisRega replied:
@sylvestre

sylvestre commented on src/uucore/src/lib/lib.rs:
```diff
@@ -84,3 +168,83 @@ pub fn args() -> impl Iterator<Item = String> {
+        );
+    }
+
+    fn test_invalid_utf8_args_ignore(os_str: &OsStr) {
```
ditto?

ChrisRega replied:
Same here: Called from line 248 in lib.rs allows for testing broken encoding on different platforms.
